### PR TITLE
Add implicit `safeFree` for `RapidsBuffer`

### LINKED ISSUE #5098: [BUG] Harden calls to `RapidsBuffer.free`
As @gerashegalov mentioned in https://github.com/NVIDIA/spark-rapids/pull/5093#pullrequestreview-926548173, there are various places in the code where `free` is called on a `RapidsBuffer` instance with the assumption that it does not throw.  We either should have something like a `safeFree` implicit utility function or appropriate `try...finally` blocks to ensure attempts to release other resources at the same time are not interrupted.

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi [13716567376yh@gmail.com](mailto:13716567376yh@gmail.com)

Close #5098

Add implicit classes to safely free RapidsBuffers

### CODE REVIEW

**Comment by HaoYang670** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala [hunk 1]
```diff
@@ -59,6 +59,30 @@ object RapidsPluginImplicits {
...
+      if (rapidsBuffer != null) {
+        if (e != null) {
+          try {
+            rapidsBuffer.free()
+          } catch {
+            case suppressed: Throwable => e.addSuppressed(suppressed)
+          }
+        } else {
+          rapidsBuffer.free()
+        }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala:81]:
```suggestion
        try {
          rapidsBuffer.free()
        } catch {
          case suppressed: Throwable if e != null => e.addSuppressed(suppressed)
        }
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala [hunk 2]
```diff
@@ -87,11 +111,44 @@ object RapidsPluginImplicits {
...
+        throw freeException
+      }
+    }
+  }
+
   implicit class AutoCloseableArray[A <: AutoCloseable](val in: Array[A]) {
     def safeClose(e: Throwable = null): Unit = if (in != null) {
       in.toSeq.safeClose(e)
     }
   }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/implicits.scala:146]:
```suggestion
  }

```

**Comment by HaoYang670** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**